internal/db/db.go

```diff
@@ -49,57 +61,116 @@ func (d *Database) Transaction(f func(*Database) error) error {
-// then the migration will be performed no matter what the current version
-// is. The force parameter should only be set when the migration is
-// initiated by a user request.
-func (d *Database) Migrate(ctx context.Context, force bool) error {
+// by the current data model. If the database is not configured then an error
+// with a code of errors.CodeServerConfiguration will be returned.
+func (d *Database) Migrate(ctx context.Context) error {
 	const op = errors.Op("db.Migrate")
 	if d == nil || d.DB == nil {
 		return errors.E(op, errors.CodeServerConfiguration, "database not configured")
 	}
 	db := d.DB.WithContext(ctx)
-	schema, _ := dbmodel.SQL.ReadFile(path.Join("sql", db.Name(), "versions.sql"))
-	if err := db.Exec(string(schema)).Error; err != nil {
-		return errors.E(op, dbError(err))
+
+	sqlDB, err := db.DB()
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to obtain raw DB: %w", err))
+	}
+	conn, err := sqlDB.Conn(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to obtain DB conn: %w", err)
 	}
 
-	for {
-		v := dbmodel.Version{Component: dbmodel.Component, Major: 1, Minor: 0}
-		if err := db.FirstOrCreate(&v).Error; err != nil {
-			return errors.E(op, dbError(err))
-		}
-		if dbmodel.Major == v.Major && dbmodel.Minor <= v.Minor {
-			// The database is already at, or past, our current version.
-			// Nothing to do.
-			atomic.StoreUint32(&d.migrated, 1)
-			return nil
-		}
-		if v.Major != dbmodel.Major && !force {
-			return errors.E(op, errors.CodeServerConfiguration, fmt.Sprintf("database has incompatible version %d.%d", v.Major, v.Minor))
+	sqlDir, err := iofs.New(dbmodel.SQL, path.Join("sql", db.Name()))
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to create new sql filesys: %w", err))
+	}
+
+	driver, err := postgres.WithConnection(ctx, conn, &postgres.Config{MigrationsTable: MigrationTableName})
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to create new driver instance: %w", err))
+	}
+
+	// DB name is left blank because it is contained in the driver/DB connection.
+	m, err := migrate.NewWithInstance("iofs", sqlDir, "", driver)
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to create new migrator: %w", err))
+	}
+	defer m.Close()
+
+	// Setup custom logger for consistent output.
+	logger := migrationLogger{logger: zapctx.Logger(ctx), verbose: false}
+	m.Log = logger
+
+	if err := d.handleDeprecatedMigrations(ctx, m); err != nil {
+		return errors.E(op, fmt.Errorf("failed to handle deprecated migrations: %w", err))
+	}
+
+	v, dirty, err := m.Version()
+	if err != nil {
+		if !stderr.Is(err, migrate.ErrNilVersion) {
+			return errors.E(op, fmt.Errorf("failed to get db version: %w", err))
 		}
-		// The major versions are unchanged, the database can be migrated.
-		v.Minor += 1
-		schema, err := dbmodel.SQL.ReadFile(path.Join("sql", db.Name(), fmt.Sprintf("%d_%d.sql", v.Major, v.Minor)))
-		if err != nil {
-			return errors.E(op, err)
+	}
+
+	if dirty {
```

Comment:
just to make sure I understand. If the migration we've started before fails, it sets the dirty flag in the db, and the next time we call migrate it forces the version to the one before it and reruns the migration.
Right?

Reply:
Correct!